src/Slices/ServiceInstanceDetails/Test/AttributesUtils.test.ts

```diff
@@ -0,0 +1,407 @@
+        site: {
+          name: "inmanta-lab",
+          infra_vendor: "vcenter",
+        },
+        system_id: 1,
+        epc_version: "11.3.2-1",
+        telco_ip_range: "192.168.126.0/24",
+      },
+    });
+  });
+
+  it("should return all attribute sets if they all exist in the selected log", () => {
+    const logs: InstanceLog[] = historyData;
+
+    const result = getAvailableAttributesSets(logs, "5");
+
+    expect(result).toEqual({
+      active_attributes: {
+        name: "core1",
+        site: {
+          name: "inmanta-lab",
+          infra_vendor: "vcenter",
+        },
+        system_id: 1,
+        epc_version: "11.3.2-1",
+        telco_ip_range: "192.168.126.0/24",
+      },
+      candidate_attributes: {
+        name: "core1",
+        site: {
+          name: "inmanta-lab",
+          infra_vendor: "vcenter",
+        },
+        system_id: 1,
+        epc_version: "11.3.2-1",
+        telco_ip_range: "192.168.126.0/24",
+      },
+      rollback_attributes: {
+        name: "core1",
+        site: {
+          name: "inmanta-lab",
+          infra_vendor: "vcenter",
+        },
+        system_id: 1,
+        epc_version: "11.3.2-1",
+        telco_ip_range: "192.168.126.0/24",
+      },
+    });
+  });
+});
+
+describe("formatTreeRowData", () => {
+  it("should return an empty array when the attributes object is empty", () => {
+    const result = formatTreeRowData({});
+
+    expect(result).toEqual([]);
+  });
+
+  it("should format a flat object with primitive values correctly", () => {
+    const attributes = {
+      key1: "value1",
+      key2: 123,
+      key3: true,
+    };
+
+    const result = formatTreeRowData(attributes);
+
+    expect(result).toMatchObject([
+      {
+        id: expect.stringMatching(/^key1/),
+        name: "key1",
+        value: "value1",
+        children: [],
+      },
+      {
+        id: expect.stringMatching(/^key2/),
+        name: "key2",
+        value: 123,
+        children: [],
+      },
+      {
+        id: expect.stringMatching(/^key3/),
+        name: "key3",
+        value: true,
+        children: [],
+      },
+    ]);
+  });
+
+  it("should format an object with nested objects correctly", () => {
+    const attributes = {
+      parent: {
+        child1: "value1",
+        child2: 456,
+      },
+    };
+
+    const result = formatTreeRowData(attributes);
+
+    expect(result).toMatchObject([
+      {
+        id: expect.stringMatching(/^parent/),
+        name: "parent",
+        value: { child1: "value1", child2: 456 },
+        children: [
+          {
+            id: expect.stringMatching(/^child1/),
+            name: "child1",
+            value: "value1",
+            children: [],
+          },
+          {
+            id: expect.stringMatching(/^child2/),
+            name: "child2",
+            value: 456,
+            children: [],
+          },
+        ],
+      },
+    ]);
+  });
+
+  it("should format an array of objects correctly", () => {
```

Comment:
I'm not 100% sure but maybe test scenario for nested array of array of objects would be good as well, if we are covering objects, nested objects and deeply nested objects.

Reply:
As you can see in the matchObject, this test covers deeply nested elements, where we have a list of embedded entities, which contain on themselves a extra level of nested elements